File: runtimes/dart-3.3/src/function_types.dart
import 'dart:convert';
import 'logger.dart';

class RuntimeRequest {
  List<int> bodyBinary;
  Map<String, dynamic> headers;
  String method;
  String url;
  String path;
  int port;
  String scheme;
  String host;
  String queryString;
  Map<String, String> query;

  dynamic get body {
    String contentType = (this.headers['content-type'] ?? 'text/plain').toLowerCase();

    if(contentType.startsWith("application/json")) {
      return this.bodyJson;
    }

    List<String> binaryTypes = ["application/", "audio/", "font/", "image/", "video/"];
    for(final binaryType in binaryTypes) {
      if(contentType.startsWith(binaryType)) {
        return this.bodyBinary;
      }
    }

    return this.bodyText;
  }

  String get bodyRaw {
    return this.bodyText;
  }
  
  String get bodyText {
    return utf8.decode(this.bodyBinary);
  }

  dynamic get bodyJson {
    return jsonDecode(this.bodyText);
  }

  RuntimeRequest(
      {String method = '',
      String scheme = '',
      String host = '',
      int port = 80,
      String path = '',
      Map<String, String> query = const {},
      String queryString = '',
      Map<String, dynamic> headers = const {},
      List<int> bodyBinary = const [],
      String url = '',})
      : method = method,
        scheme = scheme,
        host = host,
        port = port,
        path = path,
        query = query,
        queryString = queryString,
        headers = headers,
        bodyBinary = bodyBinary,
        url = url {}
}

class RuntimeResponse {
  dynamic binary(List<int> bytes,
      [int statusCode = 200, Map<String, dynamic> headers = const {}]) {
    return {
      'body': bytes,
      'statusCode': statusCode,
      'headers': headers,
    };
  }

  dynamic send(String body,
      [int statusCode = 200, Map<String, dynamic> headers = const {}]) {
    return this.text(body, statusCode, headers);
  }

  dynamic text(String body,
      [int statusCode = 200, Map<String, dynamic> headers = const {}]) {
    return this.binary(utf8.encode(body), statusCode, headers);
  }

  dynamic json(Map<String, dynamic> json,
      [int statusCode = 200, Map<String, dynamic> headers = const {}]) {
    var headersMerged = {...headers, 'content-type': 'application/json'};
    return this.text(jsonEncode(json), statusCode, headersMerged);
  }

  dynamic empty() {
    return this.text('', 204, const {});
  }

  dynamic redirect(String url,
      [int statusCode = 301, Map<String, dynamic> headers = const {}]) {
    var headersMerged = {...headers, 'location': url};
    return this.text('', statusCode, headersMerged);
  }
}

class RuntimeContext {
  RuntimeRequest req;
  RuntimeResponse res;
  Logger logger;

  RuntimeContext(RuntimeRequest req, RuntimeResponse res, Logger logger)
      : req = req,
        res = res,
        logger = logger {}

  void log(dynamic message) {
    this.logger.write(message, Logger.TYPE_LOG);
  }

  void error(dynamic message) {
    this.logger.write(message, Logger.TYPE_ERROR);
  }
}

File: runtimes/python/src/function_types.py
from logger import Logger
import json

class Response:
    def binary(self, bytes, statusCode = 200, headers = {}):
        return {
            'body': bytes,
            'statusCode': statusCode,
            'headers': headers,
        }
    
    def text(self, body, statusCode = 200, headers = {}):
        return self.binary(body.encode("utf-8"), statusCode, headers)

    def send(self, body, statusCode = 200, headers = {}):
        return self.text(str(body), statusCode, headers)

    def json(self, obj, statusCode = 200, headers = {}):
        headers['content-type'] = 'application/json'
        return self.text(json.dumps(obj, separators=(',', ':')), statusCode, headers)
    
    def empty(self):
        return self.text('', 204, {})

    def redirect(self, url, statusCode = 301, headers = {}):
        headers['location'] = url
        return self.text('', statusCode, headers)

class Request:
    body_binary = None
    headers = None
    method = None
    url = None
    path = None
    port = None
    query = None
    query_string = None
    scheme = None
    host = None

    @property
    def body_text(self):
        return self.body_binary.decode("utf-8")

    @property
    def body_json(self):
        return json.loads(self.body_text)

    @property
    def body_raw(self):
        return self.body_text

    @property
    def body(self):
        content_type = self.headers.get('content-type', 'text/plain').lower()

        if content_type.startswith('application/json'):
            if len(self.body_binary) > 0:
                return self.body_json
            else:
                return {}

        binary_types = ["application/", "audio/", "font/", "image/", "video/"]
        for binary_type in binary_types:
            if content_type.startswith(binary_type):
                return self.body_binary

        return self.body_text

class Context:
    req = Request()
    res = Response()
    logger = None

    logs = []
    errors = []

    def __init__(self, logger):
        self.logs = []
        self.errors = []
        self.req = Request()
        self.res = Response()
        self.logger = logger

    def log(self, message):
        self.logger.write(message, Logger.TYPE_LOG)

    def error(self, message):
        self.logger.write(message, Logger.TYPE_ERROR)
